chall06/vde-dios.c

```diff
@@ -0,0 +1,111 @@
+#include <math.h>
+#include <string.h>
+#include <stdlib.h>
+
+typedef struct s_element{
+    int x;
+    int y;
+    int pos;
+    int side;
+    char id;
+}       t_element;
+
+int ft_check_format(char *board)
+{
+    int i = -1;
+    int len = strlen(board);;
+    int side;
+
+    //Valid charset?
+    while(++i < len)
+        if (!strchr(".KPBRQ\n", board[i]))
+            return (0);
+    //Valid square?
+    if (len > 89)
+        return (0);
+    side = sqrt(len);
+    i = side;
+    while (board[i] && i < len)
+    {
+        if (board[i] != '\n')
+            return (0);
+        i += side + 1;
+    }
+    //Only one king?
+    if (strchr(strchr(board, 'K') + 1, 'K'))
+        return (0);
+    return (1);
+}
+
+t_element ft_element_coord(int pos, int side, char *board)
+{
+    t_element el;
+
+    el.x = pos / (side + 1);
+    el.y = pos - (side + 1) * el.x;
```

Comment:
Those calculation are wrong:
```
........
........
........
....B...
........
.......K
........
........
```
In this board the king is situated at `X: 7`,`Y: 5` (`ordinal: 52` in the string).  
But, given that `x = pos / (side + 1)`, this would mean that `7 = 52 / (8 + 1)` -> `7 = 52 / 9` -> `7 = 5`. Which as you already know, is wrong.
It is the same with `y = pos - (side + 1) * x` (assuming that we got `x` right this time):
`5 = 52 - (8 + 1) * 7` -> `5 = 52 - 9 * 7` -> `5 = -11`, which is also wrong.

Given this formula (which is correct): `pos = x + y * (size + 1)`, we have the following formulas:
- `x = pos - y * (size + 1)`
- `y = (pos - x) / (size +1)`

As `x` and `y` are co-dependent there are 2 solutions:
- If you want to keep going with formulas, I think there is one, but to be honest I didn't want to do more maths than necessary
- You could have incremented `x` and `y` earlier in the loop

I chose the second option, and used the first formula to compute the position in the string.

Reply:
Noo!! I made a mistake and wrote a  division `/`symbol with modulus symbol `%`. I had though about this correctly, just a transcribing issue... 

That x formula totally screws up calculations. Instead of writing this:
`x = pos / (side + 1)` 
I actually wanted to write this: 
`x = pos % (side + 1)` (which actually works like a charm)

Also I don't know how I did not figure this error out because with that division instead of modulus I did not get any errors when testing... 🤷🏻‍♂️But yes totally wrong.

So, write formulas (these ones totally work are):

- `x = pos % (side + 1)`

- `y = (pos - x) / (size +1)`